### ISSUE TITLE
Alternative representation of IxList

### ISSUE THREAD

**Issue by adamgundry** [well-typed/optics#260]
@edsko suggested we could use a redundant representation of `IxList` to avoid the problem that `NoIx` in written code turns into `'[]` in inferred types:
```hs
data IxList = NoIx | WithIx Type | Type :. IxList
```
or possibly (thanks to @kosmikus) 
```hs
data IxList = NoIx | WithIx Type | IxList :.: IxList
```

**Comment by arybczak** [comment]
Sounds like a good idea, I think that `NoIx` suddenly turning into `'[]` after composition looks a bit weird.

**Comment by arybczak** [comment]
I experimented with it a bit and it looks like it suffers from similar "tick problem" optic kinds did, i.e. you need `DataKinds` to write `'NoIx`.

**Comment by phadej** [comment]
IMO, ticks in type-errors / ghci output is not that bad we should avoid them at all cost. `doctest` outputs are tricky though.

One can argue that `'NoIx Int` is better than `'[Int]` however.

Yet, I don't like non-uniqueness of representation. Writing e.g. `Append` becomes nasty, `Curry` also. **If** (unluckily) users would like to write something own => tricky.

Thus I'm :-1: on any `IxList` which isn't isomorphic to `Nil | Cons x (IxList x)`.

**Comment by arybczak** [comment]
I found that `data IxList = NoIx | Type :. WithIx IxList` translates cleanly from the list approach (since it's isomorphic to list) and looks good, I'd be for using this form.

**Comment by adamgundry** [comment]
I've been experimenting as well. I think we can avoid the requirement for `DataKinds` or ticks on input by defining type synonyms to shadow the constructors, e.g.
```hs
type NoIx = 'NoIx
```
I think having ticks in type errors is not too bad, compared to the existing ticked lists.

A unique representation does seem better than having to define `Append` etc. for trees, and reduces the risk of nasty errors where the index-lists don't match.  My email notifications suggest @arybczak suggested
```hs
 data IxList = NoIx | Type :. WithIx IxList
```
but the comment seems to have disappeared? How is `WithIx` defined here?

**Comment by arybczak** [comment]
I don't remember the comment, but I might've written something that doesn't make sense :laughing: 

I checked both `data IxList = NoIx | WithIx Type IxList` and `data IxList = NoIx | Type :%: IxList`.

The case of one index in both of these looks weird, i.e. there is a hanging NoIx at the end which is confusing.

> type NoIx = 'NoIx

But then we're back to the same issue there was with ticked optic kinds :( If you don't export `'NoIx`, it'll be fully qualified in ghci when GHC decides you give you expanded type synonym.

After some experimentation I decided to not bother as regular lists seem alright to me given the problems of more specialized types.
